### ISSUE TITLE
Multi GPU training

### ISSUE THREAD

**Issue by gokberkgul** [MarvinLer/tcga_weakly_MIL_WSI_segmentation#10]
Hi, thanks for the awesome work.

I am trying to build my own model and I decided to use your Dataset class. However, there is one thing that I didn't understand. You have implemented the `__getitem__` function as (if we ignore `retrieve_tiles_ids_with_images` flag)

>def __getitem__(self, item):
      slide_instances = self._get_slide_instances(item)
      slide_instances = torch.stack([self.transform(instance) for instance in slide_instances])
      slide_label = self.slides_labels[item]
      return slide_instances, slide_label

So, as far as I understand, it returns `max_bag_size` number of tiles from WSI number `item`. That is why you set the batch size as 1 when loading up the data. My question is, how would that work In a multi-GPU environment? Lets say I have 4 GPU's and I want each GPU to load up 64 tiles, and I want `max_bag_size` of 256 (So that one call of dataloader would process one WSI). What should I change in the code? Or would the code work out of the box? If it was a "traditional" dataset, I would call dataloader as

`data_loader = torch.utils.data.DataLoader(
        dataset,
        sampler=sampler,
        batch_size=args.batch_size_per_gpu,
        num_workers=args.num_workers,
        pin_memory=True,
        drop_last=True,
    )`

**Comment by MarvinLer** [comment]
Hi,
Indeed this returns the number of tiles equal to the value of the `--max-bag-size` parameter. The tensor returned by the [`__getitem__`](https://github.com/MarvinLer/tcga_segmentation/blob/master/code/data_processing/pytorch_dataset.py#L173) function is of shape `[max_bag_size, 3, width, height]`. This is similar to traditional imaging datasets (e.g. imagenet) except you have an extra dimension for the bag.
It is the dataloader of pytorch that adds an extra dimension such that the shape returned by the dataloader is `[1, max_bag_size, 3, width, height]` -- and this extra dimension has value 1 i.e. the batch size (which, for simplicity, has been fixed to 1 in this code although it is generalizable to higher batch sizes).

So if you just use a higher batch size `b` in the dataloader (i.e. sample from `b` WSI), the returned shape will be `[b, max_bag_size, 3, width, height]` which is what you want I believe. 
If you want to have a total of `max_bag_size` sampled tiles, you should actually divide by the number of gpus.

In other words, the pytorch dataloader calls `batch_size` times the `__getitem__` of a dataset, and stacks the results with the [`default_collate`](https://github.com/pytorch/pytorch/blob/master/torch/utils/data/_utils/collate.py#L43) function.

Please tell me if you have any issue
ML

**Comment by gokberkgul** [comment]
My aim is actually to get `[max_bag_size/n_gpus, 3, width, height]` per GPU so that if I use `n_gpus`, all of them can process one WSI at the same time. I will try to change the line to (same for slide_labels)
`slide_instances = self._get_slide_instances(int(item/n_gpus))`
And make `_get_slide_instances` return `max_bag_size/n_gpus` number of tiles, I think that would fix my problem right?

**Comment by MarvinLer** [comment]
`item` is actually an index: the dataset is supposed to seek the data associated with the sample of index `item`; this is on par with the way the pytorch `Dataset` is implemented.

What you want can be done without changing anything about the [`Dataset`](https://github.com/MarvinLer/tcga_segmentation/blob/master/code/data_processing/pytorch_dataset.py#L22) class with two steps:
1. put your number of gpus in the `batch_size` parameter in your dataloader that wraps the `Dataset`
2. change the command line argument `--max-bag-size` (or the parameter `max_bag_size` of [`Dataset`](https://github.com/MarvinLer/tcga_segmentation/blob/master/code/data_processing/pytorch_dataset.py#L22)) to the number of tiles that you want per gpu

If you denote `n` the desired number of tiles per gpu, you will get a tensor size `[1, n, 3, width, height]` per gpu (which you can `squeeze(0)` to get rid of the first dimension).

Is it okay for you this way?
ML
